prediction_market_agent/agents/metaculus_agent/deploy.py

```diff
@@ -1,84 +1,83 @@
         )
-        logger.info(f"Found {len(markets)} open markets to submit predictions for.")
 
-        if not self.repeat_predictions:
-            # Filter out markets that we have already answered
-            markets = [market for market in markets if not market.have_predicted]
-            logger.info(
-                f"Found {len(markets)} unanswered markets to submit predictions for."
+    def get_markets(self, market_type: MarketType) -> Sequence[AgentMarket]:  # type: ignore # TODO: Needs to be decided in https://github.com/gnosis/prediction-market-agent/pull/511#discussion_r1810034688 and then I'll implement it here.
+        markets: Sequence[MetaculusAgentMarket] = (
+            MetaculusAgentMarket.get_binary_markets(
+                limit=self.bet_on_n_markets_per_run,
+                tournament_id=self.tournament_id,
+                filter_by=FilterBy.OPEN,
+                sort_by=SortBy.NEWEST,
             )
+        )
+        return markets
+
+    def verify_market(self, market_type: MarketType, market: AgentMarket) -> bool:
+        assert isinstance(
+            market, MetaculusAgentMarket
+        ), "Just making mypy happy. It's true thanks to the check in the `run` method via `supported_markets`."
 
-        for market in markets:
-            logger.info(f"Answering market {market.id}, question: {market.question}")
-            if not self.dummy_prediction:
-                # TODO incorporate 'Resolution criteria', 'Fine print', and
-                # 'Background info' into the prompt given to the agent.
-                answer = agent.answer_binary_market(
-                    market.question, created_time=market.created_time
-                )
-            else:
-                answer = ProbabilisticAnswer(
-                    p_yes=Probability(0.5),
-                    reasoning="Just a test.",
-                    confidence=0.5,
-                )
+        # Filter out the market if the agent isn't configured to re-bet.
+        if not self.repeat_predictions and market.have_predicted:
+            return False
 
-            if answer is None:
-                logger.error("No answer was given. Skipping")
-            else:
-                market.submit_prediction(
-                    p_yes=answer.p_yes,
-                    reasoning=check_not_none(answer.reasoning),
-                )
+        # Otherwise all markets on Metaculus are fine.
+        return True
+
+    def answer_binary_market(self, market: AgentMarket) -> ProbabilisticAnswer | None:
+        assert isinstance(
+            market, MetaculusAgentMarket
+        ), "Just making mypy happy. It's true thanks to the check in the `run` method via `supported_markets`."
+        logger.info(f"Answering market {market.id}, question: {market.question}")
+        answer: ProbabilisticAnswer | None
+        if not self.dummy_prediction:
+            full_question = f"""Question: {market.question}
```

Comment:
The previous Metaculus tournament wasn't tracked in Langfuse, so there is not much to debug.

Thanks to this refactoring it will be tracked now, example: https://cloud.langfuse.com/project/clv2hvvyw0006z9uchz6u1irw/traces/15a2dd9d-20e7-4ef3-9235-c41050064661?observation=3c1861b6-8866-4e27-8182-4196d091e08e

I thought this is a fine way to add additional fields to the existing agents -- if it turns out that it works, we could make it officially like this. Something like the `full_question` property on AgentMarket.

Reply:
🤞